test/test_public_bindings.py

```diff
@@ -272,6 +275,65 @@ def test_no_new_bindings(self):
         msg = f"torch._C had bindings that are not present in the allowlist:\n{difference}"
         self.assertTrue(torch_C_bindings.issubset(torch_C_allowlist_superset), msg)
 
+    def test_correct_module_names(self):
+        '''
+        An API is considered public, if  its  `__module__` starts with `torch.`
+        and there is no name in `__module__` or the object itself that starts with “_”.
+        Each public package should either:
+        - (preferred) Define `__all__` and all callables and classes in there must have their
+         `__module__` start with the current submodule's path. Things not in `__all__` should
+          NOT have their `__module__` start with the current submodule.
+        - (for simple python-only modules) Not define `__all__` and all the elements in `dir(submod)` must have their
+          `__module__` that start with the current submodule.
+        '''
+        failure_list = []
+
+        def test_module(modname):
+            is_a_private_module = False
+            split_strs = modname.split('.')
+            mod = sys.modules.get(modname)
+            for elem in split_strs:
+                if elem.startswith("_"):
+                    is_a_private_module = True
+                    break
+
+            if is_a_private_module:
+                return
+
+            # verifies that each API has the correct module name and naming semantics
+            # depending on whether it's public or private
+            def verify_correct_modulename(elem, modname, mod, private_api):
+                obj = getattr(mod, elem)
+                if not (isinstance(obj, Callable) or inspect.isclass(obj)):
+                    return
+                elem_module = getattr(obj, '__module__', None)
+                elem_modname_starts_with_modname = elem_module is not None and elem_module.startswith(modname)
+                if private_api:
+                    # elem's name must begin with an `_` and it's module name
+                    # should NOT start with it's current module since it's a private API
+                    if not elem.startswith('_') or elem_modname_starts_with_modname:
```

Comment:
why isn't this a "and"?
Having only one of the two doesn't make that API "public"

Reply:
If a class name or function begins with an underscore then it's public no (we have already checked above that the module it is in is public)?

Similarly if an API is not public, then it's module name should not begin with current the current module.

Reply:
I'm not sure this work for both.

> If a class name or function begins with an underscore then it's public no

I assume you mean "not public" right?
If so, then the answer is yes for the dir() case but not for the `__all__` case.

Reply:
I am not sure I follow your point